carlog/carlog.go

```diff
@@ -1623,6 +1625,27 @@ func (j *CarLog) Close() error {
 	j.pendingReads.Wait() // writes hold idxLk
 
 	if ri == nil {
+		if j.data != nil {
+			_, err := j.Commit()
+			if err != nil {
+				return xerrors.Errorf("committing head: %w", err)
+			}
+			if err := j.data.Close(); err != nil {
+				return xerrors.Errorf("closing data: %w", err)
+			}
+		}
+
+		if j.head != nil {
+			if err := j.head.Close(); err != nil {
+				return xerrors.Errorf("closing head: %w", err)
+			}
+		}
+
+		if j.eIdx != nil {
+			if err := j.eIdx.Close(); err != nil {
+				return xerrors.Errorf("closing e index: %w", err)
+			}
+		}
```

Comment:
Oh, this is a great catch, we should eventually (not in this PR) add a test which checks that no FDs are leaking (kind of like what we have in lotus e.g. [here](https://github.com/filecoin-project/lotus/blob/3f93b86ebea667cd94685f84db0dfa54ac5c5ead/storage/sealer/ffiwrapper/sealer_test.go#L597-L641))